File: jpeg-client/cpp/jpeg-client.cpp
// Copyright (C) 2023 KOSEKI Yoshinori
///
/// @file  jpege-client.cpp
/// @brief JPEGサーバ
/// @author KOSEKI Yoshinori <koseki.y@gmail.com>
///

#include <iostream>
#include <zmq.hpp>
#include <opencv2/opencv.hpp>

#include "jpeg-client.hpp"

// コンストラクタ
JPEGClient::JPEGClient (const std::string port)
{
  context = new zmq::context_t(1);
  socket = new zmq::socket_t(*context, ZMQ_REQ);
  socket->connect ("tcp://localhost:" + port);

  std::cerr << "Connecting to JPEG server…" << std::endl;
}

// デストラクタ
JPEGClient::~JPEGClient()
{
}

// 1フレーム受信
void JPEGClient::run(cv::Mat& image_proc)
{
  { // request
    std::string request = "send JPEG";

    int size = request.length();
    zmq::message_t message(size);
    memcpy(message.data(), request.c_str(), size);
    socket->send(message);
  }

  // get jpeg
  {
    zmq::message_t message;

    // recieve: data
    socket->recv(&message);

    int size = message.size();
    std::vector<unsigned char> buff(size);
    memcpy(&buff[0], (unsigned char*)message.data(), size);
    image_proc = cv::imdecode(cv::Mat(buff), 1); // CV_LOAD_IMAGE_COLOR
  }
}

/// Local Variables: ///
/// truncate-lines:t ///
/// End: ///

File: jpeg-client/python/jpeg-client.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import cv2 as cv
import numpy as np
import zmq

class JPEGClient():

    def __init__(self, port="5555"):

        context = zmq.Context()
        self.socket = context.socket(zmq.REQ)
        self.socket.connect("tcp://localhost:" + port)

        print("Connecting to JPEG server…")


    def recieve(self):

        self.socket.send(b"send JPEG")  # リクエスト

        message = self.socket.recv()                       # 受信
        data = np.frombuffer(message, dtype=np.uint8) # numpy配列へ変換
        image = cv.imdecode(data, 3)                  # decode JPEG

        if 1 < len(message):
            return True, image

        return False, []


def main():

    client = JPEGClient()

    while True:
        ret, image_in = client.recieve()
        if not ret:
            break

        cv.imshow("jpeg client (python)", image_in)

        key = cv.waitKey(30)

        if key == 27:  # ESC
            break


if __name__ == '__main__':
    main()

# import pdb; pdb.set_trace()

### Local Variables: ###
### truncate-lines:t ###
### End: ###
